File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        d = dict()

        # Initializatin
        prefixSum = 0
        count = 0
        d[0]=1

        # Traverse
        for i in range(n):
            prefixSum += nums[i]

            if (prefixSum-k) in d:
                count += d[prefixSum-k]


            if prefixSum not in d:
                d[prefixSum] = 1
            else:
                d[prefixSum] += 1

        # Return
        return count
        
        

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int,int> mp;

        //Initialization
        int prefixSum = 0;
        int count = 0;
        mp[0] = 1;

        //Traverse
        for(int i=0; i<n; i++){
            prefixSum += nums[i];

            if(mp.find(prefixSum-k) != mp.end()){
                count += mp[prefixSum-k];
            }

            mp[prefixSum]++;
        }
        
        return count;
    }
};